examples/simple/simple.cpp

```diff
@@ -114,24 +112,25 @@ int main(int argc, char ** argv) {
 
     while (n_cur <= n_predict) {
         // sample the next token
+        llama_token new_token_id = llama_sampler_sample(smpl, ctx, -1);
         {
-            const llama_token new_token_id = llama_sampler_sample(smpl, ctx, -1);
 
             // is it an end of generation?
             if (llama_token_is_eog(model, new_token_id) || n_cur == n_predict) {
-                LOG("\n");
+                fprintf(stderr, "\n");
 
                 break;
             }
 
-            LOG("%s", llama_token_to_piece(ctx, new_token_id).c_str());
+            char buf[128];
+            int n = llama_token_to_piece(model, new_token_id, buf, sizeof(buf), 0, true);
+            std::string s(buf, n);
+            fprintf(stderr, "%s", s.c_str());
```

Comment:
I recently found that some models can have tokens with length of up to 256. So it would be a good idea to check for negative `n` here.

Btw, is `std::string` necessary here? can we just set null terminator manually and print `buf` directly?

The logic for the `LOG_` messages was to output the generated text to `stdout` and everything else to `stderr`. This has the benefit of being able to redirect info messages to somewhere else (e.g. `/dev/null`).

Reply:
I do not really understand why `llama_token_to_piece` does not return a NUL-terminated string in the first place. I assume that it is because some tokens may contain `\0` characters, in which case it is necessary to use `std::string` to represent them, even though they are discarded again in the call to `printf`.

Printing the generated text to `stderr` was a mistake, I will fix that.